File: win-prog/ProgWinVB/Taming the Mouse/EnterLeave/EnterLeave.vb
'-------------------------------------------
' EnterLeave.vb (c) 2002 by Charles Petzold
'-------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class EnterLeave
    Inherits Form

    Private bInside As Boolean = False

    Shared Sub Main()
        Application.Run(New EnterLeave())
    End Sub

    Sub New()
        Text = "Enter/Leave"
    End Sub

    Protected Overrides Sub OnMouseEnter(ByVal ea As EventArgs)
        bInside = True
        Invalidate()
    End Sub

    Protected Overrides Sub OnMouseLeave(ByVal ea As EventArgs)
        bInside = False
        Invalidate()
    End Sub

    Protected Overrides Sub OnMouseHover(ByVal ea As EventArgs)
        Dim grfx As Graphics = CreateGraphics()
        grfx.Clear(Color.Red)
        System.Threading.Thread.Sleep(100)
        grfx.Clear(Color.Green)
        grfx.Dispose()
    End Sub

    Protected Overrides Sub OnPaint(ByVal pea As PaintEventArgs)
        Dim grfx As Graphics = pea.Graphics

        If bInside Then
            grfx.Clear(Color.Green)
        Else
            grfx.Clear(BackColor)
        End If
    End Sub
End Class

File: win-prog/ProgWinCS/Taming the Mouse/EnterLeave/EnterLeave.cs
//-----------------------------------------
// EnterLeave.cs © 2001 by Charles Petzold
//-----------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class EnterLeave: Form
{
     bool bInside = false;

     public static void Main()
     {
          Application.Run(new EnterLeave());
     }
     public EnterLeave()
     {
          Text = "Enter/Leave";
     }
     protected override void OnMouseEnter(EventArgs ea)
     {
          bInside = true;
          Invalidate();
     }
     protected override void OnMouseLeave(EventArgs ea)
     {
          bInside = false;
          Invalidate();
     }
     protected override void OnMouseHover(EventArgs ea)
     {
          Graphics grfx = CreateGraphics();
    
          grfx.Clear(Color.Red);
          System.Threading.Thread.Sleep(100);
          grfx.Clear(Color.Green);
          grfx.Dispose();
     }
     protected override void OnPaint(PaintEventArgs pea)
     {
          Graphics grfx = pea.Graphics;

          grfx.Clear(bInside ? Color.Green : BackColor);
     }
}
